pandas/core/sparse/array.py

```diff
@@ -120,319 +152,464 @@ def _wrap_result(name, data, sparse_index, fill_value, dtype=None):
 
-    @property
-    def shape(self):
-        return (len(self),)
+        elif method is not None:
+            warnings.warn("Converting to dense in fillna with 'method'",
+                          PerformanceWarning)
+            filled = interpolate_2d(np.asarray(self), method=method,
+                                    limit=limit)
+            return type(self)(filled, fill_value=self.fill_value)
 
-    @property
-    def sp_values(self):
-        # caching not an option, leaks memory
-        return self.view(np.ndarray)
+        else:
+            new_values = np.where(isna(self.sp_values), value, self.sp_values)
 
-    @property
-    def fill_value(self):
-        return self._fill_value
+            if self._null_fill_value:
+                # This is essentially just updating the dtype.
+                new_dtype = SparseDtype(self.dtype.subtype, fill_value=value)
+            else:
+                new_dtype = self.dtype
 
-    @fill_value.setter
-    def fill_value(self, value):
-        if not is_scalar(value):
-            raise ValueError('fill_value must be a scalar')
-        # if the specified value triggers type promotion, raise ValueError
-        new_dtype, fill_value = maybe_promote(self.dtype, value)
-        if is_dtype_equal(self.dtype, new_dtype):
-            self._fill_value = fill_value
+        return self._simple_new(new_values, self._sparse_index, new_dtype)
+
+    def shift(self, periods=1):
+
+        if periods == 0:
+            return self.copy()
+
+        subtype = np.result_type(np.nan, self.dtype.subtype)
+
+        if subtype != self.dtype.subtype:
+            # just coerce up front
+            arr = self.astype(SparseDtype(subtype, self.fill_value))
         else:
-            msg = 'unable to set fill_value {fill} to {dtype} dtype'
-            raise ValueError(msg.format(fill=value, dtype=self.dtype))
+            arr = self
 
-    def get_values(self, fill=None):
-        """ return a dense representation """
-        return self.to_dense(fill=fill)
+        empty = self._from_sequence([self.dtype.na_value] * abs(periods),
+                                    dtype=arr.dtype)
+        if periods > 0:
+            a = empty
+            b = arr[:-periods]
+        else:
+            a = arr[abs(periods):]
+            b = empty
+        return arr._concat_same_type([a, b])
 
-    def to_dense(self, fill=None):
+    def _first_fill_value_loc(self):
         """
-        Convert SparseArray to a NumPy array.
+        Get the location of the first missing value.
+
+        Returns
+        -------
+        int
+        """
+        if len(self) == 0 or self.sp_index.npoints == len(self):
+            return -1
+
+        indices = self.sp_index.to_int_index().indices
+        if indices[0] > 0:
+            return 0
+
+        diff = indices[1:] - indices[:-1]
+        return np.searchsorted(diff, 2) + 1
+
+    def unique(self):
```

Comment:
I've ignored docstrings that are inherited, when the parent seems fine. For example, I wrote a new fillna docstring since the performance needed to be explained.

Of the following, do you prefer

1. Inheriting the docstring as is
2. Inheriting the doctring and doing a string substation of ExtensionArray for `cls.__name__`
3. Manually re-writing the docstring

Reply:
I am fine with inheriting the docstring as is, for this PR, as that seems the simplest (to not further complicate this PR). 
We can then later think how to do this (we will have the same problem for the other arrays)